File: src/LabSync.Core/Entities/Job.cs
using LabSync.Core.Types;

namespace LabSync.Core.Entities;

public class Job
{
    public Guid Id { get; init; }

    public Guid DeviceId { get; init; }
    public Device? Device { get; private set; }

    public string Command { get; init; }
    public string Arguments { get; init; }
    public string? ScriptPayload { get; init; }

    public JobStatus Status { get; private set; }
    public int? ExitCode { get; private set; }
    public string? Output { get; private set; }

    public DateTime CreatedAt { get; init; }
    public DateTime? FinishedAt { get; private set; }

    protected Job() { }

    public Job(Guid deviceId, string command, string arguments = "", string? scriptPayload = null)
    {
        Id = Guid.NewGuid();
        DeviceId      = deviceId;
        Command       = command;
        Arguments     = arguments;
        ScriptPayload = scriptPayload;
        Status        = JobStatus.Pending;
        CreatedAt     = DateTime.UtcNow;
    }

    public void MarkAsRunning()
    {
        if (Status != JobStatus.Pending)
            throw new InvalidOperationException("Only pending tasks can be started.");

        Status = JobStatus.Running;
    }

    public void Complete(int exitCode, string? output)
    {
        if (Status != JobStatus.Running)
            throw new InvalidOperationException("You can only complete tasks that are currently running.");

        ExitCode   = exitCode;
        Output     = output;
        Status     = exitCode == 0 ? JobStatus.Completed : JobStatus.Failed;
        FinishedAt = DateTime.UtcNow;
    }

    public void Cancel()
    {
        if (Status == JobStatus.Completed || Status == JobStatus.Failed)
            return;

        Status     = JobStatus.Cancelled;
        FinishedAt = DateTime.UtcNow;
    }
}

File: src/client/src/types/job.ts
export const JobStatus = {
  Pending: 0,
  Running: 1,
  Completed: 2,
  Failed: 3,
  Cancelled: 4,
} as const;

export type JobStatus = (typeof JobStatus)[keyof typeof JobStatus];

export interface JobDto {
  id: string;
  deviceId: string;
  command: string;
  arguments: string;
  status: JobStatus;
  exitCode: number | null;
  output: string | null;
  createdAt: string;
  finishedAt: string | null;
}

/**
 * Mirrors LabSync.Core.Dto.CreateJobRequest
 */
export interface CreateJobRequest {
  command: string;
  arguments: string;
  scriptPayload?: string | null;
}

export const JOB_STATUS_LABELS: Record<JobStatus, string> = {
  [JobStatus.Pending]: 'Pending',
  [JobStatus.Running]: 'Running',
  [JobStatus.Completed]: 'Completed',
  [JobStatus.Failed]: 'Failed',
  [JobStatus.Cancelled]: 'Cancelled',
};
